### PR TITLE
GSAGH-497 - GsaGH creates releases opposite to what is defined using 'Edit element/member' component

### PR DESCRIPTION
**Pull request by SandeepArup**

### CODE REVIEW

File: GsaGH/Helpers/GH/ReplaceParam.cs
Comment (psarras) [inline GsaGH/Helpers/GH/ReplaceParam.cs:1]: why we need this update?

File: GsaGH/Helpers/StringExtension.cs [hunk 1]
```diff
@@ -38,5 +42,109 @@ public static string TrimSpaces(this string value) {
...
+        } else if (mystring == "fix" || mystring == "fixed") {
+          bool6.X = !isRelease;
+          bool6.Y = !isRelease;
+          bool6.Z = !isRelease;
+          bool6.Xx = !isRelease;
+          bool6.Yy = !isRelease;
+          bool6.Zz = !isRelease;
+          return bool6;
+        } else if (mystring == "release" || mystring == "released" || mystring == "hinge"
+          || mystring == "hinged" || mystring == "charnier") {
```
Comment (psarras) [inline GsaGH/Helpers/StringExtension.cs:92]: wrap with a method or two. IsHinged(), IsReleased()

File: GsaGH/Helpers/StringExtension.cs [hunk 2]
```diff
@@ -38,5 +42,109 @@ public static string TrimSpaces(this string value) {
...
+          || FindString(booleanString, "yy☐") || FindString(booleanString, "yy✓")
+          || FindString(booleanString, "zz☐") || FindString(booleanString, "zz✓")) {
+
+        input.X = FindString(booleanString, "x✓");
+        input.Y = FindString(booleanString, "y✓");
+        input.Z = FindString(booleanString, "z✓");
+        input.Xx = FindString(booleanString, "xx✓");
+        input.Yy = FindString(booleanString, "yy✓");
+        input.Zz = FindString(booleanString, "zz✓");
+        return true;
```
Comment (psarras) [inline GsaGH/Helpers/StringExtension.cs:124]: hmm why do we check this? Can we not check the True / False values instead?

File: GsaGH/Helpers/StringExtension.cs [hunk 3]
```diff
@@ -38,5 +42,109 @@ public static string TrimSpaces(this string value) {
...
+        Xx = ConvertChar(txt[i++], isRelease),
+        Yy = ConvertChar(txt[i++], isRelease),
+        Zz = ConvertChar(txt[i++], isRelease),
+      };
+      return b;
+    }
+    private static bool ConvertChar(char character, bool isRelease) {
+      return character switch {
+        'r' => isRelease,
+        'f' => !isRelease,
```
Comment (psarras) [inline GsaGH/Helpers/StringExtension.cs:144]: do we also need the capital versions of those?

File: GsaGH/Parameters/1_Properties/GsaReleaseParameter.cs
Comment (psarras) [inline GsaGH/Parameters/1_Properties/GsaReleaseParameter.cs:1]: Did we not said to create a common ancestor for GsaReleaseParameter and GsaRestrainParameter?

**Review by psarras — CHANGES_REQUESTED** [review]

File: GsaGH/Helpers/GH/ReplaceParam.cs
Comment (SandeepArup) [reply GsaGH/Helpers/GH/ReplaceParam.cs:1]: Update is required as opening file will have info of GsaBool6Parameter but that does not know how to parse "FFFRRR" as those cast has been removed from GsaBool6Parameter.

File: GsaGH/Parameters/1_Properties/GsaReleaseParameter.cs
Comment (SandeepArup) [reply GsaGH/Parameters/1_Properties/GsaReleaseParameter.cs:1]: Is this what you are expecting i.e. Parameter inherited from GsaBool6Goo? https://github.com/arup-group/GSA-Grasshopper/pull/721/files#diff-2146e40799484c09b42c0927d6f7fcb4a9490500c84cc3c0b360a3c019ecd183

File: GsaGH/Helpers/GH/ReplaceParam.cs [hunk 4]
```diff
@@ -36,5 +38,63 @@ public static string UnsupportedValue(GH_ObjectWrapper ghTypeWrapper) {
       type = type.Replace("Goo", string.Empty);
       return type;
     }
+
+    public static void UpdateReleaseBool6Parameter(this GH_ComponentParamServer parameters) {
```
Comment (DominikaLos) [inline GsaGH/Helpers/GH/ReplaceParam.cs:42]: why not 2 methods? one for input and one for output

File: GsaGH/Helpers/GH/ReplaceParam.cs [hunk 5]
```diff
@@ -36,5 +38,63 @@ public static string UnsupportedValue(GH_ObjectWrapper ghTypeWrapper) {
...
+            NickName = param.NickName,
+            Description = param.Description,
+            Access = param.Access
+          };
+          parameters.ReplaceOutputParameter(parameter, id);
+        }
+      }
+    }
+
+    public static void UpdateRestrainedBool6Parameter(this GH_ComponentParamServer parameters) {
```
Comment (DominikaLos) [inline GsaGH/Helpers/GH/ReplaceParam.cs:71]: as above

File: GsaGHTests/1_BaseParameters/1_Properties/GsaBool6Test.cs [hunk 6]
```diff
@@ -83,5 +88,67 @@ public void DuplicateTest(bool x, bool y, bool z, bool xx, bool yy, bool zz) {
       Assert.Equal(yy, original.Yy);
       Assert.Equal(zz, original.Zz);
     }
+
+    [Fact]
+    public void ReleaseParameterInfoIsCorrect() {
```
Comment (DominikaLos) [inline GsaGHTests/1_BaseParameters/1_Properties/GsaBool6Test.cs:93]: split to include one assert per test

File: GsaGHTests/2_GooWrappers/GsaBool6GooTest.cs [hunk 7]
```diff
@@ -1,59 +1,144 @@
...
+        Z = expectedZ,
+        Xx = expectedXx,
+        Yy = expectedYy,
+        Zz = expectedZz
+      };
+      Assert.Equal(releaseBool6, output);
+
+      //restraint is the opposite of release
+      output = StringExtension.ParseRestrain(text);
+      Assert.Equal(!releaseBool6, output);
```
Comment (DominikaLos) [inline GsaGHTests/2_GooWrappers/GsaBool6GooTest.cs:112]: as above

File: GsaGHTests/3_Components/2_Geometry/EditElement1dTests.cs [hunk 8]
```diff
@@ -30,6 +30,17 @@ public static GH_OasysComponent ComponentMother() {
       return comp;
     }
 
+    internal static void CompareRelease(GsaBool6 input, GsaBool6 output) {
```
Comment (DominikaLos) [inline GsaGHTests/3_Components/2_Geometry/EditElement1dTests.cs:33]: as above

File: GsaGHTests/1_BaseParameters/1_Properties/GsaBool6Test.cs (hunk 6 above)
Comment (DominikaLos) [inline GsaGHTests/1_BaseParameters/1_Properties/GsaBool6Test.cs:93]: one assert per test

File: GsaGHTests/3_Components/2_Geometry/EditElement1dTests.cs [hunk 9]
```diff
@@ -30,6 +30,17 @@ public static GH_OasysComponent ComponentMother() {
...
+      Assert.Equal(input.ToString(), output.ToString());
+      Assert.Equal(input.ToString(), output.ToString());
+      Assert.Equal(input.X, output.X);
```
Comment (DominikaLos) [inline GsaGHTests/3_Components/2_Geometry/EditElement1dTests.cs:36]: as above

File: GsaGH/Components/2_Geometry/EditNode.cs [hunk 10]
```diff
@@ -59,7 +59,7 @@ public override bool Read(GH_IReader reader) {
         Params.ReplaceInputParameter(new GsaSpringPropertyParameter(), 7, true);
         Params.ReplaceOutputParameter(new GsaSpringPropertyParameter(), 7);
       }
-
+      Params.UpdateRestrainedBool6Parameter();
```
Comment (psarras) [inline GsaGH/Components/2_Geometry/EditNode.cs:62]: is this for upgrading existing components with Bool6Params?

**Review by psarras — APPROVED** [review]

File: GsaGH/Components/2_Geometry/EditNode.cs (hunk 10 above)
Comment (SandeepArup) [reply GsaGH/Components/2_Geometry/EditNode.cs:62]: Yes....No release are set by ReleaseParameter and restrained by RestraintParameter. So, It is essential to update

File: GsaGH/Helpers/GH/ReplaceParam.cs (hunk 4 above)
Comment (SandeepArup) [reply GsaGH/Helpers/GH/ReplaceParam.cs:42]: @DominikaLos @psarras I prefer least call. Single function call here will have least chance of error like updating input but   missing to update output parameter. However, I am open for your suggestion.Kinldy confirm again and I will update code.

File: GsaGHTests/1_BaseParameters/1_Properties/GsaBool6Test.cs (hunk 6 above)
Comment (SandeepArup) [reply GsaGHTests/1_BaseParameters/1_Properties/GsaBool6Test.cs:93]: @psarras @DominikaLos   I think when checking one output against one input then one assert make sense. Here single creation of object return think which we want to test. Checking name, description, nickname in separate test will make  it lengthy and multiple call of same thing.

File: GsaGHTests/2_GooWrappers/GsaBool6GooTest.cs (hunk 7 above)
Comment (SandeepArup) [reply GsaGHTests/2_GooWrappers/GsaBool6GooTest.cs:112]: @DominikaLos @psarras  Release/restraint is property of type bool6. It is single data class which hold 6 plain simple boolean data. I do not see reason why we should check x,y,z, xx, yy and zz in separate test

File: GsaGHTests/3_Components/2_Geometry/EditElement1dTests.cs [hunk 11]
```diff
@@ -289,5 +289,31 @@ public void InvalidPropertyElementTypeCombination3() {
...
+      GH_OasysComponent comp = ComponentMother();
+      var startReleaseInput = new GsaBool6(true, true, true, false, false, false);
+      var endReleaseInput = new GsaBool6(false, false, false, true, true, true);
+      ComponentTestHelper.SetInput(comp, new GsaBool6Goo(startReleaseInput), 7);
+      ComponentTestHelper.SetInput(comp, new GsaBool6Goo(endReleaseInput), 8);
+      GsaBool6 startReleaseOutput = ((GsaBool6Goo)ComponentTestHelper.GetOutput(comp, 7)).Value;
+      GsaBool6 endReleaseOutput = ((GsaBool6Goo)ComponentTestHelper.GetOutput(comp, 8)).Value;
+      Assert.Equal(startReleaseInput, startReleaseOutput);
+      Assert.Equal(endReleaseInput, endReleaseOutput);
+    }
```
Comment (psarras) [inline GsaGHTests/3_Components/2_Geometry/EditElement1dTests.cs:304]: can you test its input in a separate test, no need to compare them at the same time

File: GsaGHTests/3_Components/2_Geometry/EditElement1dTests.cs [hunk 12]
```diff
@@ -289,5 +289,31 @@ public void InvalidPropertyElementTypeCombination3() {
...
+      GH_OasysComponent comp = ComponentMother();
+      var startReleaseInput = new GsaBool6(true, true, true, false, false, false);
+      var endReleaseInput = new GsaBool6(false, false, false, true, true, true);
+      ComponentTestHelper.SetInput(comp, "RRRFFF", 7);
+      ComponentTestHelper.SetInput(comp, "FFFRRR", 8);
+      GsaBool6 startReleaseOutput = ((GsaBool6Goo)ComponentTestHelper.GetOutput(comp, 7)).Value;
+      GsaBool6 endReleaseOutput = ((GsaBool6Goo)ComponentTestHelper.GetOutput(comp, 8)).Value;
+     Assert.Equal(startReleaseInput, startReleaseOutput);
+     Assert.Equal(endReleaseInput, endReleaseOutput);
+    }
```
Comment (psarras) [inline GsaGHTests/3_Components/2_Geometry/EditElement1dTests.cs:317]: same here

File: GsaGHTests/3_Components/2_Geometry/EditMember1dTests.cs [hunk 13]
```diff
@@ -349,5 +349,31 @@ public void InvalidPropertyElement1DTypeCombination3() {
...
+      GH_OasysComponent comp = ComponentMother();
+      var startReleaseInput = new GsaBool6(true, true, true, false, false, false);
+      var endReleaseInput = new GsaBool6(false, false, false, true, true, true);
+      ComponentTestHelper.SetInput(comp, "RRRFFF", 8);
+      ComponentTestHelper.SetInput(comp, "FFFRRR", 9);
+      GsaBool6 startReleaseOutput = ((GsaBool6Goo)ComponentTestHelper.GetOutput(comp, 8)).Value;
+      GsaBool6 endReleaseOutput = ((GsaBool6Goo)ComponentTestHelper.GetOutput(comp, 9)).Value;
+      Assert.Equal(startReleaseInput, startReleaseOutput);
+      Assert.Equal(endReleaseInput, endReleaseOutput);
+    }
```
Comment (psarras) [inline GsaGHTests/3_Components/2_Geometry/EditMember1dTests.cs:377]: same here

File: GsaGHTests/2_GooWrappers/GsaBool6GooTest.cs [hunk 14]
```diff
@@ -1,59 +1,144 @@
...
+    public void Bool6ShouldThrowExceptionWhen6Characters() {
+      string exactlySixCharacters = "ABCDEF";
+      Assert.Throws<InvalidCastException>(() => StringExtension.ParseBool6(exactlySixCharacters));
+    }
+
+    [Fact]
+    public void ParseReleaseShouldThrowExceptionWhenANonStringNonBool() {
+      var exactlySixCharacters = new DummyClass();
+      Assert.Throws<InvalidCastException>(() => StringExtension.ParseRelease(exactlySixCharacters));
+    }
```
Comment (psarras) [inline GsaGHTests/2_GooWrappers/GsaBool6GooTest.cs:140]: one of them meant to capture this untested code line 123:

image

Could you try and understand what we need to do to get it?

Comment (SandeepArup) [reply GsaGHTests/2_GooWrappers/GsaBool6GooTest.cs]: I could not reach to that point except when null data is passed. Updated test for null

File: GsaGHTests/3_Components/2_Geometry/EditElement1dTests.cs (hunk 12 above)
Comment (SandeepArup) [reply GsaGHTests/3_Components/2_Geometry/EditElement1dTests.cs]: @psarras all done. Kindly have a look

**Review by psarras — DISMISSED** [review]

File: GsaGH/Helpers/StringExtension.cs [hunk 15]
```diff
@@ -36,7 +43,129 @@ public static string TrimSpaces(this string value) {
       while (value.Contains("  ")) {
         value = value.Replace("  ", " ");
       }
+
       return value.Trim();
     }
+
+    public static GsaBool6 ParseBool6(object data) {
```
Comment (DominikaLos) [inline GsaGH/Helpers/StringExtension.cs:50]: why it is in string extension class?

File: GsaGH/Helpers/StringExtension.cs [hunk 16]
```diff
@@ -36,7 +43,129 @@ public static string TrimSpaces(this string value) {
...
+
+    private static string ExceptionMessage(object data) {
+      if (data != null) {
+        return $"Data  conversion failed from '{data}' of type {data.GetTypeName()} to Bool6";
+      } else {
+        return $"Data  conversion failed from null to Bool6";
+      }
+    }
+
+    public static GsaBool6 ParseRestrain(object data) {
```
Comment (DominikaLos) [inline GsaGH/Helpers/StringExtension.cs:74]: as above

File: GsaGH/Helpers/StringExtension.cs [hunk 17]
```diff
@@ -36,7 +43,129 @@ public static string TrimSpaces(this string value) {
...
+        } else {
+          throw new InvalidCastException(ExceptionMessage(data));
+        }
+
+        return bool6;
+      }
+      throw new InvalidCastException(ExceptionMessage(data));
+    }
+
+    private static bool IsRelease(string input) {
```
Comment (DominikaLos) [inline GsaGH/Helpers/StringExtension.cs:133]: as above. 
restrain and release shouldn't be here

**Review by DominikaLos — CHANGES_REQUESTED** [review]

File: GsaGH/Helpers/StringExtension.cs (hunk 17 above)
Comment (SandeepArup) [reply GsaGH/Helpers/StringExtension.cs]: @DominikaLos have moved this into different class and .cs file

File: GsaGH/Helpers/StringExtension.cs [hunk 18]
```diff
@@ -1,6 +1,13 @@
-﻿using System.Linq;
+﻿using System;
+using System.Linq;
 using System.Text.RegularExpressions;
 
+using Grasshopper.Kernel;
+
+using GsaGH.Parameters;
+
+using Rhino.Render.Fields;
```
Comment (DominikaLos) [inline GsaGH/Helpers/StringExtension.cs:9]: why new namespaces if there is no change?

**Review by DominikaLos — APPROVED** [review]

File: GsaGH/Helpers/StringExtension.cs (hunk 18 above)
Comment (SandeepArup) [reply GsaGH/Helpers/StringExtension.cs]: @DominikaLos Good point. Have removed.

File: GsaGH/Parameters/1_Properties/GsaReleaseParameter.cs [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+      Category = CategoryName.Name();
+      SubCategory = SubCategoryName.Cat9();
+    }
+
+    protected override GsaBool6Goo PreferredCast(object data) {
+      try {
+        GsaBool6 gsaBool6 = ParseBool6.ParseRelease(data);
+        return new GsaBool6Goo(gsaBool6);
+      } catch (InvalidCastException e) {
+        AddRuntimeMessage(GH_RuntimeMessageLevel.Error, e.Message);
```
Comment (DominikaLos) [inline GsaGH/Parameters/1_Properties/GsaReleaseParameter.cs:37]: use AddRuntimeError method

File: GsaGH/Parameters/1_Properties/GsaRestrainedParameter.cs [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+      Category = CategoryName.Name();
+      SubCategory = SubCategoryName.Cat9();
+    }
+
+    protected override GsaBool6Goo PreferredCast(object data) {
+      try {
+        GsaBool6 gsaBool6 = ParseBool6.ParseRestrain(data);
+        return new GsaBool6Goo(gsaBool6);
+      } catch (InvalidCastException e) {
+        AddRuntimeMessage(GH_RuntimeMessageLevel.Error, e.Message);
```
Comment (DominikaLos) [inline GsaGH/Parameters/1_Properties/GsaRestrainedParameter.cs:36]: as above

**Review by DominikaLos — APPROVED** [review]

**Pull request merged**